api/cloud/oc_cloud_apis.c

```diff
@@ -267,10 +267,29 @@ cloud_deregister(cloud_api_param_t *p)
 
   OC_DBG("try deregister");
   bool cannotConnect = true;
+  // This value is calculated by coap_oscore_serialize_message for deregister
+  // message with empty query parameters. The value should remain the same
+  // unless some global options are added to coap requests.
+  // The deregister request won't be sent if the total size of its header is
+  // greater than COAP_MAX_HEADER_SIZE, so we must ensure that the query
+  // is not too large.
+  // Some older cloud implementations require tokens in deregister requests.
+  // To facilitate support for such implementations we append access token
+  // to the request query if the resulting query size is within the limit.
+  #define DEREGISTER_EMPTY_QUERY_HEADER_SIZE 38
+  oc_string_t query = cloud_access_deregister_query(oc_string(ctx->store.uid),
+    oc_string(ctx->store.access_token), /*device*/0);
+  size_t query_size = oc_string_len(query);
+  oc_free_string(&query);
+  const char* access_token = NULL;
+  if (DEREGISTER_EMPTY_QUERY_HEADER_SIZE + query_size <= COAP_MAX_HEADER_SIZE) {
+    access_token = oc_string(ctx->store.access_token);
+  }
   if (oc_string(ctx->store.ci_server) && conv_cloud_endpoint(ctx) == 0 &&
       cloud_access_deregister(ctx->cloud_ep,
                               oc_string(ctx->store.uid),
-                              0,
+                              access_token,
+                              /*device*/0,
```

Comment:
why there is 0 instead of the device?

Reply:
Not really sure, but looking in my original commit ( https://github.com/iotivity/iotivity-lite/pull/38/commits/c91cba16ae488be908d0217cd667e5a3f36649f6 ) it seems that the value for device has been always zero.

Reply:
could you change it to the device?

Reply:
Which value shoud I use though, ctx->store.device?

Reply:
yes